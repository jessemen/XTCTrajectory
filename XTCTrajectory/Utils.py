#-------------------------------------------------------------------------------
# . File      : Utils.py
# . Program   : pDynamo-1.8.0                           (http://www.pdynamo.org)
# . Copyright : CEA, CNRS, Martin  J. Field  (2007-2012),
#                          Mikolaj J. Feliks (2014)
# . License   : CeCILL French Free Software License     (http://www.cecill.info)
#-------------------------------------------------------------------------------
"""Utility functions for the XTCTrajectory module."""

from pBabel                  import SystemGeometryTrajectory
from XTCTrajectoryFileReader import XTCTrajectoryFileReader
from XTCTrajectoryFileWriter import XTCTrajectoryFileWriter


def XTCTrajectory_ToSystemGeometryTrajectory (inPath, outPath, system):
    """Convert an XTC trajectory to a SystemGeometryTrajectory."""
    inTrajectory  = XTCTrajectoryFileReader  (inPath,  system)
    outTrajectory = SystemGeometryTrajectory (outPath, system, mode="w")

    inTrajectory.ReadHeader ()
    while inTrajectory.RestoreOwnerData (): outTrajectory.WriteOwnerData ()

    inTrajectory.Close  ()
    outTrajectory.Close ()


def XTCTrajectory_FromSystemGeometryTrajectory (outPath, inPath, system):
    """Convert a SystemGeometryTrajectory to an XTC trajectory."""
    inTrajectory  = SystemGeometryTrajectory (inPath,  system, mode="r")
    outTrajectory = XTCTrajectoryFileWriter  (outPath, system)

    outTrajectory.WriteHeader ()
    while inTrajectory.RestoreOwnerData (): outTrajectory.WriteOwnerData ()

    inTrajectory.Close  ()
    outTrajectory.Close ()


#===============================================================================
# Testing
#===============================================================================
if __name__ == "__main__": pass
