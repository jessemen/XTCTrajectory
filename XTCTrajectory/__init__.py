#-------------------------------------------------------------------------------
# . File      : __init__.py
# . Program   : pDynamo-1.8.0                           (http://www.pdynamo.org)
# . Copyright : CEA, CNRS, Martin  J. Field  (2007-2012),
#                          Mikolaj J. Feliks (2014)
# . License   : CeCILL French Free Software License     (http://www.cecill.info)
#-------------------------------------------------------------------------------
"""A module for reading trajectories from XTC files of Gromacs."""

