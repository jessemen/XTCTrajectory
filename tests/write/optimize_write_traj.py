# Example script for the XTCTrajectory module

from pCore             import logFile
from pBabel            import CHARMMParameterFiles_ToParameters, CHARMMPSFFile_ToSystem, XYZFile_ToCoordinates3, XYZFile_FromSystem, SystemGeometryTrajectory
from pMolecule         import NBModelABFS
from pMoleculeScripts  import ConjugateGradientMinimize_SystemGeometry
from XTCTrajectory     import XTCTrajectoryFileWriter, XTCTrajectory_FromSystemGeometryTrajectory

import os


#===========================================================
logFile.Header ("Infrared fluorescent protein (parent)")


parameters = (
"par_all27_prot_na.inp",
"par.inp",
)

mol = CHARMMPSFFile_ToSystem ("parent_waterbox.psf", isXPLOR = True, parameters = CHARMMParameterFiles_ToParameters (parameters))

mol.coordinates3 = XYZFile_ToCoordinates3 ("geometry.xyz")

mol.Summary ()

nb_model = NBModelABFS ()

mol.DefineNBModel (nb_model)


#===========================================================
traj  = XTCTrajectoryFileWriter  ("optimize.xtc",  mol)

# This will write 189 MB of PKL files
# traj2 = SystemGeometryTrajectory ("optimize_traj", mol, mode = "w")


# ConjugateGradientMinimize_SystemGeometry (mol, logFrequency = 1, maximumIterations = 100, rmsGradientTolerance = 0.04, trajectories = [(traj, 1), (traj2, 1)])

ConjugateGradientMinimize_SystemGeometry (mol, logFrequency = 1, maximumIterations = 100, rmsGradientTolerance = 0.04, trajectories = [(traj, 1)])


#===========================================================
logFile.Footer ()
