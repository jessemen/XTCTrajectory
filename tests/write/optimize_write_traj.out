--------------------------------------------------------------------------------
                      Infrared fluorescent protein (parent)
--------------------------------------------------------------------------------

Start Time: Thu Sep 18 18:42:22 2014

----------------------------------------------------------------------------------------------------
                                      Text File Reader Warnings
----------------------------------------------------------------------------------------------------
   Line                                        Message                                     Severity
----------------------------------------------------------------------------------------------------
3724                                   Unable to process non-epsilon/sigma nonbond for FE.   Warning
----------------------------------------------------------------------------------------------------

There have been warnings. Proceed with caution!


---------------------- CHARMM Parameter File Line Summary ----------------------
Angle                  =            624  Bond                   =            251
Dihedral               =           1054  HBond                  =              1
Improper               =             74  NonBond                =            131
Title                  =             18
--------------------------------------------------------------------------------

-------------------- CHARMM Parameter File Parameter Summary -------------------
Atom Types             =            131  Angles                 =            623
Bonds                  =            250  Cmaps                  =              0
Dihedrals (Nonwild)    =            791  Dihedrals (Wild)       =             92
Impropers (Nonwild)    =             48  Impropers (Wild)       =             25
Nonbonds (Nonwild)     =            129  Nonbonds (Wild)        =              0
Nonbonds 1-4 (Nonwild) =             19  Nonbonds 1-4 (Wild)    =              0
Urey-Bradleys          =            145
--------------------------------------------------------------------------------

---------------------- CHARMM Parameter File Line Summary ----------------------
Angle                  =             60  Bond                   =             23
Dihedral               =            123  Improper               =             21
NonBond                =              8  Title                  =              2
--------------------------------------------------------------------------------

-------------------- CHARMM Parameter File Parameter Summary -------------------
Atom Types             =             28  Angles                 =             59
Bonds                  =             22  Cmaps                  =              0
Dihedrals (Nonwild)    =            107  Dihedrals (Wild)       =              0
Impropers (Nonwild)    =             18  Impropers (Wild)       =              2
Nonbonds (Nonwild)     =              7  Nonbonds (Wild)        =              0
Nonbonds 1-4 (Nonwild) =              0  Nonbonds 1-4 (Wild)    =              0
Urey-Bradleys          =              3
--------------------------------------------------------------------------------

---------------------------- CHARMM PSF File Summary ---------------------------
Acceptors              =              0  Angles                 =          17581
Atoms                  =          30463  Bonds                  =          22014
Dihedrals              =          13335  Donors                 =              0
Exclusions             =              0  Groups                 =              1
Impropers              =            870  ST2s                   =              0
Title                  =             11
--------------------------------------------------------------------------------

--------------------------------------------------------------------------------
Summary for System "REMARKS original generated structure x-plor psf file REMARKS
--------------------------------------------------------------------------------

---------------------------- Atom Container Summary ----------------------------
Number of Atoms        =          30463  Number of Heavy Atoms  =          10988
Number of Hydrogens    =          19475  Number of Unknowns     =              0
--------------------------------------------------------------------------------

----------------------------- Connectivity Summary -----------------------------
Atoms                  =          30463
--------------------------------------------------------------------------------

--------------------------------------------------------------------------------
                        Summary for Energy Model "CHARMM"
--------------------------------------------------------------------------------

----------------------------------- Summary for MM Model "CHARMM" ----------------------------------
LJ 1-4 Scaling                   =          1.000  El. 1-4 Scaling                  =          1.000
Number of MM Atoms               =          30463  Number of MM Atom Types          =             52
Total MM Charge                  =          -0.00  Harmonic Bond Terms              =          22014
Harmonic Bond Parameters         =            101  Harmonic Angle Terms             =          17581
Harmonic Angle Parameters        =            229  Urey-Bradley Terms               =           4161
Urey-Bradley Parameters          =             56  Fourier Dihedral Terms           =          14064
Fourier Dihedral Parameters      =            342  Harmonic Improper Terms          =            870
Harmonic Improper Parameters     =             38  Exclusions                       =          52619
1-4 Interactions                 =          13024  LJ Parameters Form               =          AMBER
LJ Parameters Types              =             52  1-4 LJ Parameters Form           =          AMBER
1-4 LJ Parameters Types          =             52
----------------------------------------------------------------------------------------------------

------------------------------- Sequence Summary -------------------------------
Number of Atoms            =      30463  Number of Components       =       8838
Number of Entities         =          8  Number of Linear Polymers  =          0
Number of Links            =          0  Number of Variants         =          0
--------------------------------------------------------------------------------

-------------------------------  Conjugate Gradient Minimizer Options ------------------------------
Beta Type                        =              1  Initial Step                     =            0.1
Line Searcher                    =           None  Log Frequency                    =              1
Maximum Iterations               =            100  Maximum Theta                    =          1e+10
Minimum Theta                    =          1e-10  RMS Gradient Tolerance           =           0.04
Steepest Descent Tolerance       =          0.001  Use Spectral Theta               =           True
----------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------------------------------------
  Iteration       Function          RMS Gradient        Max. |Grad.|          RMS Disp.         Max. |Disp.|
----------------------------------------------------------------------------------------------------------------
     0     I    -302723.07812116         70.45328642        732.91250951          0.00033079          0.00344115
     1   L4s    -333469.18503474         22.26828878        265.45249174          0.00955755          0.10652005
     2   L1s    -344176.67798496         31.03044147       1018.37444561          0.01103108          0.13591364
     3   L0s    -352491.60252182         21.58048154        579.46156479          0.01103108          0.13404085
     4   L1s    -356818.84856673         19.37241320        416.64030370          0.00729501          0.08997118
     5   L1s    -367307.00579080         28.50949611       1385.07707418          0.02114842          0.23475545
     6   L1s    -372996.95507626         17.29167696        444.07781046          0.01156339          0.14590605
     7   L1s    -376623.01971354         14.38241608        427.10390843          0.00684588          0.09050589
     8   L0s    -379215.24884970         15.24302249        765.17167544          0.00684588          0.09595167
     9   L1s    -382177.04187677         17.02069846        873.00001646          0.00878548          0.11534888
    10   L1s    -386106.18245763         17.52213390        587.93163822          0.01232480          0.15683703
    11   L0s    -389515.76821495         15.63287011        754.53943665          0.01232480          0.14959735
    12   L1s    -391844.81213662         12.92510261        419.42630312          0.00899919          0.09790362
    13   L1s    -393754.95990097         11.96989687        331.65048435          0.00745028          0.08280672
    14   L0s    -395640.01080021         11.29208514        394.56653947          0.00745028          0.07990529
    15   L1s    -397545.29368011         13.04564666        455.65840632          0.00836976          0.08650362
    16   L1s    -399934.99541072         13.20313404        420.75318206          0.01176893          0.11485663
    17   L1s    -401921.76693269         11.38041861        346.32689774          0.01032897          0.10407779
    18   L1s    -403407.79753705         10.43159363        591.46121943          0.00794985          0.10660341
    19   L1s    -404621.07963126          9.57000791        553.56103848          0.00690028          0.07517947
    20   L1s    -405968.34908224         10.83765999        434.22917054          0.00830559          0.09431765
    21   L1s    -407474.67050786         10.79500656        332.42263181          0.00974586          0.11489910
    22   L0s    -409047.51519218         10.19174619        348.85112674          0.00974586          0.11800081
    23   L1s    -410427.36400985          9.70137005        314.79964328          0.00879994          0.10906412
    24   L0s    -411610.60723370          9.70298322        307.69746365          0.00879994          0.10791577
    25   L0s    -412697.31640400          9.70319364        317.81538593          0.00879994          0.09827129
    26   L0s    -413777.80587479          9.41253385        356.88063174          0.00879994          0.09431942
    27   L1s    -415002.21010615          9.71554643        317.63456757          0.00982226          0.11967118
    28   L0s    -416106.16298894         10.20599351        408.20052406          0.00982226          0.11927115
    29   L0s    -417160.37790567          8.80965252        256.58625621          0.00982226          0.11187406
    30   L1s    -418073.76331416          8.40340699        262.46482631          0.00774088          0.08752960
    31   L1s    -419108.64355576          8.93988710        306.07346804          0.00880796          0.09917902
    32   L0s    -420142.44156405          8.53409256        281.95091361          0.00880796          0.09092420
    33   L0s    -421101.45171562          8.68389256        324.91093522          0.00880796          0.09650560
    34   L0s    -421986.71254316          8.31154334        331.80756740          0.00880796          0.10656161
    35   L0s    -422920.23921124          8.23728163        229.82767458          0.00880796          0.11067643
    36   L0s    -423893.97361392          7.89964017        228.74687505          0.00880796          0.10358838
    37   L0s    -424737.43661273          8.67257699        278.66143516          0.00880796          0.08670966
    38   L0s    -425548.68962876          7.85450147        245.61406389          0.00880796          0.07882024
    39   L1s    -426267.39595571          7.61661264        253.25371047          0.00728498          0.07195725
    40   L1s    -427129.75376306          8.37265821        257.32616212          0.00914588          0.08920198
    41   L0s    -427959.83805765          7.96036859        243.23350138          0.00914588          0.07480624
    42   L0s    -428733.02668038          8.00491456        330.61419087          0.00914588          0.06501585
    43   L1s    -429400.70514288          6.96046461        183.06676020          0.00784542          0.06626454
    44   L0s    -430103.70572894          7.67213011        234.87159002          0.00784542          0.06723958
    45   L1s    -430875.57514659          7.49200968        314.11931791          0.00889522          0.06572176
    46   L1s    -431555.13797693          7.04190317        236.00484234          0.00805280          0.06190330
    47   L0s    -432206.68264137          7.08309045        244.15022797          0.00805280          0.06143389
    48   L0s    -432812.52329198          7.41516683        283.88741800          0.00805280          0.06882542
    49   L0s    -433457.08899422          6.85416837        256.59893851          0.00805280          0.06861756
    50   L0s    -434068.58116943          7.40882755        294.66352589          0.00805280          0.06943404
    51   L0s    -434655.45895261          6.51977995        227.61980347          0.00805280          0.06211666
    52   L1s    -435207.05949394          6.85650449        257.66588168          0.00702363          0.05345292
    53   L1s    -435818.01196531          6.76847780        245.63296076          0.00795776          0.06194419
    54   L0s    -436391.80953761          7.08708051        235.20566761          0.00795776          0.06663949
    55   L0s    -436982.36217389          6.32168535        186.47609199          0.00795776          0.07152900
    56   L1s    -437513.38058858          6.33298802        198.15256856          0.00698067          0.06456209
    57   L1s    -438074.92283140          6.69075279        238.00486459          0.00781028          0.07095544
    58   L0s    -438610.54509744          6.62283767        244.35696656          0.00781028          0.06717203
    59   L0s    -439149.39353233          6.18778561        250.12026500          0.00781028          0.06749394
    60   L1s    -439603.30373644          5.87118517        219.11930973          0.00654109          0.05594769
    61   L1s    -440093.51964686          6.31283250        272.10192413          0.00738047          0.06038635
    62   L0s    -440586.49473338          6.01723203        257.94093149          0.00738047          0.05907580
    63   L0s    -441042.63167166          6.07685768        265.84042068          0.00738047          0.06094861
    64   L0s    -441444.95714376          6.20366205        281.32069073          0.00738047          0.06383891
    65   L0s    -441878.15455797          5.65991718        222.34549391          0.00738047          0.06849770
    66   L1s    -442297.02371575          5.64870463        226.41625918          0.00655966          0.06151373
    67   L0s    -442716.19303653          5.61882877        253.02852480          0.00655966          0.05678987
    68   L0s    -443129.13479214          5.61215410        255.99199782          0.00655966          0.05552848
    69   L1s    -443584.67055886          5.82460775        269.15643984          0.00761423          0.06553002
    70   L1s    -443982.73257247          5.57850978        285.87262431          0.00682672          0.05997460
    71   L0s    -444381.20217090          5.45041137        280.54237278          0.00682672          0.06166616
    72   L0s    -444750.32617485          5.63495895        264.58910189          0.00682672          0.06193621
    73   L1s    -445155.08277434          5.56360732        245.47812299          0.00766371          0.07023741
    74   L1s    -445527.68054439          5.17288871        242.09207275          0.00676277          0.06260118
    75   L0s    -445872.56361491          5.81450183        310.13861501          0.00676277          0.06315109
    76   L0s    -446237.02878147          4.84682761        190.53357952          0.00676277          0.06187844
    77   L1s    -446566.03017302          5.17893892        213.78824661          0.00556476          0.04949106
    78   L1s    -446963.49696460          5.62451872        265.37610452          0.00715359          0.06241914
    79   L0s    -447354.15789590          5.46808143        254.12830440          0.00715359          0.06047419
    80   L0s    -447706.67148139          5.40152266        257.67518093          0.00715359          0.05781590
    81   L1s    -448005.94844264          4.88432579        211.30893222          0.00608433          0.05197557
    82   L0s    -448332.31512804          5.05169033        218.29677193          0.00608433          0.05528132
    83   L0s    -448673.73535820          4.85674804        215.32465803          0.00608433          0.05242208
    84   L0s    -449009.99935793          5.16606063        258.84557237          0.00608433          0.05118885
    85   L0s    -449349.42138581          4.60407897        204.39055980          0.00608433          0.05719766
    86   L0s    -449657.61598896          5.36739317        269.50386205          0.00608433          0.05925621
    87   L1s    -449988.80099516          5.09520809        232.22457312          0.00687812          0.07079311
    88   L1s    -450287.83791911          4.63435162        171.05476405          0.00583671          0.06342687
    89   L0s    -450576.60549785          4.84657900        194.95608473          0.00583671          0.05937634
    90   L0s    -450875.17921240          4.60675953        194.08984636          0.00583671          0.05712033
    91   L0s    -451155.27591135          5.00331794        225.83651468          0.00583671          0.04794718
    92   L0s    -451437.99974042          4.34831084        190.79866254          0.00583671          0.04977572
    93   L1s    -451705.60284007          4.59716036        232.07101856          0.00512344          0.04491202
    94   L0s    -451980.48336474          4.40010589        190.85861030          0.00512344          0.04517449
    95   L0s    -452267.65452394          4.22913445        168.53881903          0.00512344          0.04480871
    96   L1s    -452553.59255060          4.72946587        224.50513038          0.00574487          0.04972749
    97   L0s    -452829.27977381          4.41921847        204.05117464          0.00574487          0.04915726
    98   L0s    -453078.53061857          4.72513789        235.51142110          0.00574487          0.04965003
    99   L0s    -453322.91674590          4.25796518        210.50765322          0.00574487          0.04943876
   100   L1s    -453545.91598095          4.04349890        197.98369227          0.00484510          0.04123243
----------------------------------------------------------------------------------------------------------------

Too many iterations.

-------------------  Conjugate Gradient Minimizer Statistics -------------------
Iterations             =            100  Function Calls         =            149
--------------------------------------------------------------------------------

CPU Time:  1m 52.219s

Stop Time: Thu Sep 18 18:44:15 2014
